### PR TITLE
[Bugfix] Fix Qwen3.5 GatedDeltaNet in_proj_ba Marlin failure at TP>=2

### PR DESCRIPTION
**Pull request by sonusflow**
## Summary

Fixes #35924 — Qwen3.5 `GatedDeltaNet` `in_proj_ba` fails Marlin `MIN_THREAD_N=64` at TP>=2.

**Root cause:** `in_proj_ba` uses `MergedColumnParallelLinear` with `output_sizes=[num_v_heads] * 2`. When TP-sharded, the per-rank output falls below Marlin's minimum (e.g., 397B: `64 / TP=4 = 16 < 64`).

**Fix:** Replace with two separate `ReplicatedLinear` layers (`in_proj_b`, `in_proj_a`) that preserve `quant_config`. Each rank computes the full projection and slices to its local TP partition in `forward()`.

---

## Update (2026-04-13) — Rebased onto refactored codebase

Upstream refactored GDN into a centralized `GatedDeltaNetAttention` layer in `gdn_linear_attn.py`. This PR has been fully rewritten against the new codebase. The core fix is the same: replace `MergedColumnParallelLinear` with `ReplicatedLinear` for the Qwen3.5 (non-interleaved) B/A projections. The Qwen3-Next (interleaved) path is unchanged.

---

## Changes

- **`vllm/model_executor/layers/mamba/gdn_linear_attn.py`**:
  - Add `ReplicatedLinear` import.
  - In `__init__()`: for Qwen3.5 (`gqa_interleaved_layout=False`), create `in_proj_b` and `in_proj_a` as `ReplicatedLinear` instead of a single `MergedColumnParallelLinear` `in_proj_ba`.
  - In `forward()`: for Qwen3.5 paths (both standard and LoRA), compute full replicated B/A projections and slice to local TP partition. Qwen3-Next interleaved path unchanged.
  - In `forward_xpu()`: same replicated projection + TP slicing for Qwen3.5, reconstruct concatenated BA tensor for the XPU kernel.
  - `create_ba_proj()`: now only used for Qwen3-Next (interleaved); returns single `[num_v_heads * 2]` output.

- **`vllm/model_executor/models/qwen3_5.py`**:
  - Remove `("in_proj_ba", "in_proj_b", 0)` and `("in_proj_ba", "in_proj_a", 1)` from `stacked_params_mapping` (weights load directly into separate ReplicatedLinear layers).
  - Update `packed_modules_mapping` in both `Qwen3_5ForCausalLMBase` and `Qwen3_5ForConditionalGeneration`: replace `"in_proj_ba": ["in_proj_b", "in_proj_a"]` with separate `"in_proj_b": ["in_proj_b"]` and `"in_proj_a": ["in_proj_a"]`.

## Why not `disable_tp=True` + `quant_config=None` (PR #36199)?

| Aspect | This PR | PR #36199 |
|--------|---------|-----------|
| **Quantization** | Preserved (`quant_config=quant_config`) | Dropped (`quant_config=None`) — runs FP16 |
| **packed_modules_mapping** | Updated (separate entries) | Not updated |
| **Weight loader** | Cleaned up (no stacked_params for ba) | Not touched |
| **Production tested** | 4× GB10 (SM121), 37 tok/s, days of runtime | No evidence |

Setting `quant_config=None` forces those layers to run in full precision, increasing memory usage and losing Marlin acceleration. Our approach keeps them quantized via `ReplicatedLinear`.

## Test plan

- [x] Tested on 4× NVIDIA GB10 (SM121, aarch64) with `Intel/Qwen3.5-397B-A17B-int4-AutoRound` at TP=4
- [x] Model loads correctly (199 GB, 40/40 shards, ~50 GiB/node)
- [x] Inference produces correct output (verified against bfloat16 baseline)
- [x] Performance: 37 tok/s single-user, 39 tok/s peak (matches pre-fix baseline)
- [x] CUDA graph capture completes successfully
- [x] Running in production for multiple days without issues
- [x] Rebased on current `main` (2026-04-13)
- [ ] CI tests (needs `ready` label from maintainer)

cc @Isotr0py @sighingnow @kitch2400

### CODE REVIEW

**Comment by sonusflow** [comment]
Hi @Isotr0py @kitch2400 — this PR has been fully rebased onto current `main` (the GDN refactor into `gdn_linear_attn.py`). All checks pass except CI which needs the `ready` label to run.

The fix is the same core approach: replace `MergedColumnParallelLinear` with `ReplicatedLinear` for Qwen3.5's B/A projections so they don't fall below Marlin's `MIN_THREAD_N=64` at TP>=2. Quantization is preserved (unlike `quant_config=None` approaches).

Could someone add the `ready` label when you get a chance? Happy to address any review feedback. Thanks!

**Comment by sonusflow** [comment]
Hi @tdoublep @ZJY0516 @vadiklyutiy @sighingnow — pinging per CODEOWNERS for `gdn_linear_attn.py` and `qwen3_5.py`.

This PR fixes a Marlin `MIN_THREAD_N=64` failure in Qwen3.5 `GatedDeltaNet`'s `in_proj_ba` at TP>=2 by replacing `MergedColumnParallelLinear` with two `ReplicatedLinear` layers (preserving quantization, unlike `quant_config=None` workarounds). The Qwen3-Next interleaved path is untouched.

Branch is rebased on current `main` and all non-gated checks pass. CI is blocked on the `ready` label — could one of you add it and take a look when you have a chance? Happy to address any feedback.

Tested in production on 4× GB10 with `Intel/Qwen3.5-397B-A17B-int4-AutoRound` at TP=4 (~37 tok/s, multi-day uptime). Full details in the PR description.

File: vllm/model_executor/layers/mamba/gdn_linear_attn.py [hunk 1]
```diff
@@ -320,12 +321,34 @@ def __init__(
...
+                quant_config=quant_config,
+                prefix=f"{prefix}.in_proj_b",
+            )
+            self.in_proj_a = ReplicatedLinear(
+                input_size=self.hidden_size,
+                output_size=self.num_v_heads,
+                bias=False,
+                quant_config=quant_config,
+                prefix=f"{prefix}.in_proj_a",
+            )
```
Comment (Isotr0py) [inline vllm/model_executor/layers/mamba/gdn_linear_attn.py:351]: I don't think we should replace `MergedColumnParallelLinear` for this marlin specific corner case, because this will cause performance regression from `ReplicatedLinear`.

I think the proper solution to solve this issue is adding smaller `MIN_THREAD_N` support for marlin kernels instead.

**Comment by naroam1** [comment]
Hi maintainers — chiming in as a downstream user.

We are evaluating migrating our production **Qwen3.5-27B-AWQ deployment with TP=2** to vLLM, and this PR is the only blocker keeping us from doing so. AWQ Marlin trips on `MIN_THREAD_N=64` for the `in_proj_ba` projection at TP>=2.

The author rebased on current `main` 3 weeks ago (Apr 13) and pinged for the `ready` label twice without response. Could a code-owner for `gdn_linear_attn.py` / `qwen3_5.py` take a look? Happy to validate against our production stack as soon as it lands.

CC @tdoublep @ZJY0516 @vadiklyutiy @sighingnow per the earlier ping.

Thanks for the work, @sonusflow!

**Comment by sonusflow** [comment]
Thanks @naroam1 — really appreciate the validation. This confirms it's not a corner case: your **Qwen3.5-27B-AWQ at TP=2** and our **Qwen3.5-397B-A17B-int4 at TP=4** hit the exact same `MIN_THREAD_N=64` wall via different output sizes (`num_v_heads / TP` falling below 64). Two independent production deployments on different model sizes, different SM archs, different quant schemes — same blocker.

Happy to coordinate testing once this lands.

**Comment by sonusflow** [comment]
Thanks for the review @Isotr0py — appreciate the architectural concern. A few thoughts:

**On the "performance regression" claim:** for the B/A projections specifically, `output=num_v_heads` is small (e.g. 64–128 total). With `ReplicatedLinear` each rank computes the *full* GEMM (output=64, well above `MIN_THREAD_N=64`) and slices locally — meaning Marlin actually runs efficiently on the full tile. The TP-sharded `MergedColumnParallelLinear` path produces `output_per_rank = num_v_heads/TP` (e.g. 16 at TP=4), which is exactly what fails Marlin and falls back to a much slower path. In production at TP=4 we measure 37 tok/s single-user, matching the pre-quantization bfloat16 baseline. I'd ask for benchmark evidence before accepting that ReplicatedLinear is the regression here.

**On the Marlin kernel fix:** I looked at this. `min_thread_n=64` is hardcoded in `csrc/quantization/marlin/marlin.cuh:30`, the tile configs in `marlin.cu` are all `thread_n ∈ {64,128,256}`, and `generate_kernels.py` instantiates kernels for those exact configs across every quant type and SM arch. Going below 64 means new tile configs, verifying `marlin_template.h` (~2k lines) supports `thread_n_blocks=1`, and re-validating across SM75/80/89/90/100/120. That's a substantial cross-cutting kernel PR that touches code many users depend on — and at `n=16` per rank the GEMM is small enough that kernel launch overhead likely dominates anyway, so it may not even outperform the replicated approach.

**Proposal:** land this PR to unblock the two production users above (and presumably others), and I'll file a follow-up issue tracking the proper Marlin kernel work for future improvement. Happy to address other concerns about this implementation.

**Comment by Isotr0py** [comment]
> On the "performance regression" claim: for the B/A projections specifically, output=num_v_heads is small (e.g. 64–128 total). With ReplicatedLinear each rank computes the full GEMM (output=64, well above MIN_THREAD_N=64) and slices locally — meaning Marlin actually runs efficiently on the full tile. The TP-sharded MergedColumnParallelLinear path produces output_per_rank = num_v_heads/TP (e.g. 16 at TP=4), which is exactly what fails Marlin and falls back to a much slower path. In production at TP=4 we measure 37 tok/s single-user, matching the pre-quantization bfloat16 baseline. I'd ask for benchmark evidence before accepting that ReplicatedLinear is the regression here.

Can you benchmark the performance for FP8 or unquantized models that support large `tp_size` with `ReplicatedLinear`? Marlin should be the only special case that can't work with `TP=4`.

File: vllm/model_executor/layers/mamba/gdn_linear_attn.py [hunk 2]
```diff
@@ -528,18 +547,21 @@ def forward_cuda(
         if hasattr(self, "in_proj_qkv"):
             # LoRA path (Qwen3.5 only): separate in_proj_qkv and in_proj_z
             mixed_qkv, _ = self.in_proj_qkv(hidden_states)
-            ba, _ = self.in_proj_ba(hidden_states)
```
Comment (Isotr0py) [inline vllm/model_executor/layers/mamba/gdn_linear_attn.py]: Actually, you can simply pass `disable_tp=True` in `MergedColumnParallelLinear` to disable tp and weights sharding without modifying model's architecture. Then you can disable TP only for marlin quantization models.

**Comment by sonusflow** [comment]
Good call @Isotr0py — pivoted to your `disable_tp=True` suggestion. Just pushed.

The PR is now a **single-file, +59 line change** to `gdn_linear_attn.py`:

- `create_ba_proj()` detects `GPTQMarlinConfig` / `AWQMarlinConfig` and passes `disable_tp=True` to the existing `MergedColumnParallelLinear` (no class change, no replacement).
- `forward()` and `forward_xpu()` slice `b`/`a` to the local TP partition when `disable_tp` was applied (so the downstream kernel still gets `[*, num_v_heads/tp_size]`).
- `qwen3_5.py` is reverted — no `stacked_params_mapping` or `packed_modules_mapping` changes needed since the layer name `in_proj_ba` is preserved.

This addresses your FP8/unquantized concern directly: those paths keep normal TP sharding because they skip the `disable_tp` branch. Only Marlin-backed quant gets the replicated treatment, and only because that's the kernel that requires `MIN_THREAD_N >= 64`.

Could you give it another look, and add the `ready` label so CI runs?

File: vllm/model_executor/layers/mamba/gdn_linear_attn.py [hunk 3]
```diff
@@ -614,6 +662,17 @@ def forward_xpu(
...
+            # ba_proj is replicated for Marlin; slice to local TP rank
+            # so the XPU kernel receives the expected per-rank shape.
+            num_v = self.num_v_heads
+            chunk = num_v // self.tp_size
+            start = self.tp_rank * chunk
+            b_full, a_full = projected_states_ba.chunk(2, dim=-1)
+            projected_states_ba = torch.cat(
+                [b_full[:, start:start + chunk], a_full[:, start:start + chunk]],
+                dim=-1,
+            ).contiguous()
```
Comment (Isotr0py) [inline vllm/model_executor/layers/mamba/gdn_linear_attn.py]:
```suggestion
```
Marlin linear kernel is only available on CUDA:
https://github.com/vllm-project/vllm/blob/f6537612521df1156d8ac13f524e427aca908322/vllm/model_executor/kernels/linear/mixed_precision/marlin.py#L30-L39

**Review by Isotr0py — APPROVED** [review]

**Comment by sonusflow** [comment]
Thanks @Isotr0py — really appreciate you pushing the cleanup directly, the refactor is much nicer (`maybe_disable_tp` / `split_ba` helpers, `INCConfig` coverage, `current_platform.is_cuda()` guard). I synced my local to match. Looking forward to seeing CI go green.

**Comment by sonusflow** [comment]
Heads up @Isotr0py — branch was out-of-date so I rebased onto current `main` (no conflicts, your `clean` commit preserved). The force push cleared the auto-merge you'd armed. Could you re-enable it once CI is happy? Thanks!

**Comment by sonusflow** [comment]
Hi @Isotr0py — rebased onto current `main` (was 373 commits behind, lots of GDN refactoring landed since). Fresh diff is +36/-1 in `gdn_linear_attn.py`, no other files touched.

Adjustments needed during rebase:
- `GPTQMarlinConfig` was renamed to `AutoGPTQConfig` in #38288 — import + isinstance updated accordingly.
- Added `not self.gqa_interleaved_layout` to the `maybe_disable_tp` guard. The new `fix_query_key_value_ordering` path (Qwen3-Next) uses a `mixed_ba.view(num_k_heads // tp_size, 2 * num_v_heads // num_k_heads)` reshape that assumes TP-sharded shape — replicating ba_proj there would break the view. The Marlin failure is Qwen3.5-specific anyway (output_sizes structure differs), so this narrows the workaround to exactly the broken case.
- `forward_cuda` Qwen3.5 branch routed through the new `split_ba` helper; other forward paths (ROCm/AITER, XPU, CPU) untouched since they're skipped by the `current_platform.is_cuda()` gate.

The force-push will have cleared auto-merge again — could you re-arm once CI is green? Same SQUASH commit body as before is fine.

Thanks for the patience on this one!
